src/transforms/sampler.rs

```diff
@@ -55,42 +54,53 @@ impl TransformConfig for SamplerConfig {
     }
 }
 
-#[derive(Clone, Debug)]
+#[derive(Clone)]
 pub struct Sampler {
     rate: u64,
-    key_field: String,
-    pass_list: RegexSet,
+    key_field: Option<String>,
+    exclude: Box<dyn Condition>,
+    count: u64,
 }
 
 impl Sampler {
-    pub fn new(rate: u64, key_field: Option<String>, pass_list: RegexSet) -> Self {
-        let key_field = key_field.unwrap_or_else(|| log_schema().message_key().to_string());
+    pub fn new(rate: u64, key_field: Option<String>, exclude: Box<dyn Condition>) -> Self {
         Self {
             rate,
             key_field,
-            pass_list,
+            exclude,
+            count: 0,
         }
     }
 }
 
 impl FunctionTransform for Sampler {
     fn transform(&mut self, output: &mut Vec<Event>, mut event: Event) {
-        let message = event
-            .as_log()
-            .get(&self.key_field)
-            .map(|v| v.to_string_lossy())
-            .unwrap_or_else(|| "".into());
-
         emit!(SamplerEventProcessed);
 
-        if self.pass_list.is_match(&message) {
+        if self.exclude.check(&event) {
             output.push(event);
-        } else if seahash::hash(message.as_bytes()) % self.rate == 0 {
+            return;
+        }
+
+        let value = self
+            .key_field
+            .as_ref()
+            .and_then(|key_field| event.as_log().get(key_field))
+            .map(|v| v.to_string_lossy());
```

Comment:
Would this lead to different sampling behavior across events if some do not have a value in `key_field`? I think the behavior should be consistent, but that does open the question of what to do if `key_field` is specified but empty.

Reply:
> Would this lead to different sampling behavior across events if some do not have a value in key_field

It would. But I'm assuming that we are talking about how one event having or not having a field can shift count by one, and so shift the whole sampling. We can increase count for each event so the sampling by index and by hash would be independent of each other. The rate for such events would be preserved but the sampling would be more random.

I'll add it.